### ISSUE TITLE
[FEATURE] Incorporate very-detailed type hints for traversing import hierarchy to enable ML frameworks to be the search space

### ISSUE THREAD

**Issue by SamuelMarks** [keras-team/keras#18444]
Following on from keras-team/keras-core#451 have you considered exposing the type information so that you can enumerate:

- _categorical_ All Optimizers, Losses, Layers, Metrics, &etc. available _out-of-the-box_ in JAX, PyTorch, Keras
- _continuous_ What [hyper]parameters are avaiable on each of these

So that one can optimise across these _categorical_ and _continuous_ variables to find, e.g.:

- Best performing—e.g., measured with Metric: AUCROC—Model; or ensemble thereof; on a given dataset
- Best performing Optimizer on a given series of datasets, e.g., for when you're creating a new Optimizer and want to benchmark

# Bonus use case
Extending this further, one can expand from just the out-of-the-box solutions and automatically incorporate Models, Optimizers, Losses, &etc. from new research papers into this **search-space**.

(Somewhat) related work from inside your company: 
- https://github.com/google/automl/tree/master/lion
- https://github.com/google/vizier

And from outside I want to reference my own work—publication forthcoming!—at https://github.com/SamuelMarks#multi-ml made possible with my [cdd-python](https://github.com/offscale/cdd-python)

---

Now naturally you can do this from the dynamic level (unlike my solution) or at the static level (actually modify the source code with PEP484 + later type hints). This would make the keras-team/keras-core#451 tool create a rather larger `git diff` if contributed back to the model owners; B-U-T would provide developers with a (sic. IMHO) better experience.

**Comment by SamanehSaadat** [comment]
Hi @SamuelMarks !

You can already iterate through these. For example for Optimizers you can get the list of optimizers like this:
```
optimizers = [cls for cls in keras.optimizers.Optimizer.__subclasses__()]
```
